mergedSchema.js

```diff
@@ -1,18 +1,54 @@
 import { mergeSchemas, introspectSchema, makeRemoteExecutableSchema } from "graphql-tools"
+import { ApolloLink } from "apollo-link"
 import { createHttpLink } from "apollo-link-http"
+import { setContext } from "apollo-link-context"
 import fetch from "node-fetch"
 
+import config from "./config"
 import localSchema from "./schema"
 
-export default async function mergedSchema() {
-  const convectionLink = createHttpLink({
+function createConvectionLink() {
+  const httpLink = createHttpLink({
     fetch,
     uri: process.env.CONVECTION_GRAPH_URL,
-    headers: {
-      Authorization: `Bearer ${process.env.CONVECTION_TOKEN}`,
-    },
   })
 
+  const middlewareLink = new ApolloLink((operation, forward) => forward(operation))
+
+  const authMiddleware = setContext((_request, context) => {
+    const loaders = context.graphqlContext && context.graphqlContext.res.locals.dataLoaders
+    const tokenLoader = loaders && loaders.convectionTokenLoader
+
+    // If a token loader exists for Convection (i.e. this is an authenticated request), use that token to make
+    // authenticated requests to Convection.
+    if (tokenLoader) {
+      return tokenLoader().then(({ token }) => {
+        return {
+          ...context.headers,
+          headers: {
+            authorization: `Bearer ${token}`,
+          },
+        }
+      })
+    }
+
+    // Otherwise use MP’s XApp token so that on startup MP can fetch Convection’s schema.
+    // TODO: Actually enable once Convection supports this.
+    return {
+      ...context.headers,
+      headers: {
+        // authorization: `XApp ${config.GRAVITY_XAPP_TOKEN}`,
```

Comment:
@sweir27 Does this look like what you had in mind (to allow MP to fetch Convection’s schema on startup)?

Reply:
We chatted about this - it's a hard one, maybe looking into giving Metaphysics 2 client apps, one for the APIs we forward onwards and another for the server to server access. Then the server2server one can have privileged access for getting the schema.

Reply:
Just read the discussion in #platform-humans and that makes sense to me. My only question is how easy it is for MP to identify on each incoming request if the user has access to the more privileged app role? cc @mzikherman

Reply:
But once we have that we can just load 2 stitches schemas and execute against the correct one depending on the access the authenticated user has.

Reply:
Hmm, one thing that comes to mind is that in the more privileged stitched schema we would probably have stitching code that refers to the more privileged fields and thus leak schema details through MP being OSS 🤔

One solution I can think of is to have another app that imports all of MP and then extends the schema with privileged details and make that closed source? Probably only _slightly_ more cumbersome, mostly in having to remember to add fields to the right app.

Reply:
Filed a new issue to discuss further https://github.com/artsy/metaphysics/issues/810